spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameter.scala

```diff
@@ -126,54 +141,69 @@ class AllReduceParameter[T: ClassTag](id: Long, partitionNum: Int,
     BlockManagerWrapper.putBytes(blockId, fp16param.bytes(), StorageLevel.MEMORY_ONLY_SER)
   }
 
-  def getWeightBlockId(pid : Int): BlockId = {
+  private def getWeightBlockId(pid: Int): BlockId = {
     SparkExtension.getLocalBlockId(id + "weightBytes" + pid)
   }
 
-  def getWeightPartitionId(): BlockId = {
+  private def getWeightPartitionId(): BlockId = {
     SparkExtension.getLocalBlockId(id + "weights" + partitionId)
   }
 
-  def getGradientPartitionId(): BlockId = {
+  private def getGradientPartitionId(): BlockId = {
     SparkExtension.getLocalBlockId(id + "gradients" + partitionId)
   }
 
-  def getGradientBlockId(pidFrom : Int, pidTo : Int): BlockId = {
+  private def getGradientBlockId(pidFrom: Int, pidTo: Int): BlockId = {
     SparkExtension.getLocalBlockId(id.toString + pidTo + "gradientBytes" + pidFrom)
   }
 
+  /**
+   * Use a fixed thread pool to launch a thread for each partition of the weights. Each thread
+   * requests a partition of the weights from the Spark block manager and copies it into
+   * `localParameter`.
+   *
+   * @param localParameter The Tensor that will hold the retrieved weights.
+   * @return A `FutureResult` which contains futures for each thread.
+   */
   def getWeights(localParameter: Tensor[T]): FutureResult[Int] = {
-  val bm = SparkEnv.get.blockManager
-    val tasks = (0 until partitionNum).map(pid => {
-      syncPool.submit(new Callable[Int] {
-        override def call(): Int = {
-          try {
-            val blockId = getWeightBlockId(pid)
-            val localBuffer = BlockManagerWrapper.byteBufferConvert(
-              bm.getLocalBytes(blockId).getOrElse(bm.getRemoteBytes(blockId)
-                .get))
-            val start = pid * taskSize + math.min(pid, extraSize)
-            val length = taskSize + (if (pid < extraSize) 1 else 0)
-            require(localBuffer.array().length == length * 2)
-            SerializerInstance.serialize(localBuffer).deCompress(0, localParameter, start, length)
-            BlockManagerWrapper.unlock(blockId)
-            pid
-          } catch {
-            case t : Throwable =>
-              logger.error("Error: " + ExceptionUtils.getStackTrace(t))
-              throw t
+    val bm = SparkEnv.get.blockManager
+    val tasks = (0 until partitionNum).map { pid =>
+      syncPool.submit {
+        new Callable[Int] {
+          override def call(): Int = {
+            try {
+              val blockId = getWeightBlockId(pid)
+              val localBuffer = BlockManagerWrapper.byteBufferConvert(
+                bm.getLocalBytes(blockId).getOrElse(bm.getRemoteBytes(blockId).get))
```

Comment:
BTW, this will probably throw a cryptic error message if it ever fails. It could be replaced like:

````scala
val weightBuffer = List(bm.getLocalBytes(blockId), bm.getRemoteBytes(blockId))
  .flatten.headOption.getOrElse {
    throw new RuntimeException(s"Didn't find weight block $blockId in the block manager. Did you initialize this AllReduceParameter on every executor?")
  }
val localBuffer = BlockManagerWrapper.byteBufferConvert(byteBuffer)
````